File: plugin/src/knowledge/providers/stackoverflow.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const STACKEXCHANGE_API = "https://api.stackexchange.com/2.3";

interface SearchResponse {
  items?: Question[];
}

interface Question {
  question_id: number;
  title: string;
  link: string;
  score: number;
  answer_count: number;
  is_answered: boolean;
  view_count: number;
  tags?: string[];
  owner?: {
    display_name?: string;
    reputation?: number;
  };
}

export class StackOverflowProvider implements KnowledgeProvider {
  readonly name = "stackoverflow";

  async isAvailable(): Promise<boolean> {
    return true;
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      const entries = await this.search(query, maxResults);
      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `StackOverflow lookup failed: ${message}`,
      };
    }
  }

  private formatCount(n: number): string {
    if (n >= 1_000_000) {
      return `${(n / 1_000_000).toFixed(1)}M`;
    } else if (n >= 1_000) {
      return `${(n / 1_000).toFixed(1)}k`;
    }
    return String(n);
  }

  private async search(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const params = new URLSearchParams({
      order: "desc",
      sort: "relevance",
      site: "stackoverflow",
      q: query,
      pagesize: String(limit),
      filter: "withbody",
    });

    const response = await requestUrl({
      url: `${STACKEXCHANGE_API}/search/advanced?${params}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`search failed: ${response.status}`);
    }

    const data = response.json as SearchResponse;

    return (data.items ?? []).map((q) => {
      const views = this.formatCount(q.view_count);
      const answeredIcon = q.is_answered ? "[+]" : "[-]";

      const lines: string[] = [];
      lines.push(`Score: ${q.score} | Answers: ${q.answer_count} ${answeredIcon} | Views: ${views}`);
      if (q.tags && q.tags.length > 0) {
        lines.push(`Tags: ${q.tags.slice(0, 5).join(", ")}`);
      }
      if (q.owner?.display_name) {
        const rep = q.owner.reputation ? ` (${this.formatCount(q.owner.reputation)})` : "";
        lines.push(`Asked by: ${q.owner.display_name}${rep}`);
      }

      return {
        title: q.title,
        summary: lines.join("\n"),
        url: q.link,
        source: "stackoverflow",
        metadata: {
          questionId: q.question_id,
          score: q.score,
          answerCount: q.answer_count,
          isAnswered: q.is_answered,
          viewCount: q.view_count,
          tags: q.tags,
        },
      };
    });
  }
}

File: crates/mcp/src/knowledge/stackoverflow.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const STACKEXCHANGE_API: &str = "https://api.stackexchange.com/2.3";

pub struct StackOverflowProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
struct SearchResponse {
    items: Option<Vec<Question>>,
}

#[derive(Debug, Deserialize)]
struct Question {
    question_id: u64,
    title: String,
    link: String,
    score: i32,
    answer_count: u32,
    is_answered: bool,
    view_count: u32,
    tags: Option<Vec<String>>,
    owner: Option<Owner>,
}

#[derive(Debug, Deserialize)]
struct Owner {
    display_name: Option<String>,
    reputation: Option<u32>,
}

impl StackOverflowProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }

    fn format_count(n: u32) -> String {
        if n >= 1_000_000 {
            format!("{:.1}M", n as f64 / 1_000_000.0)
        } else if n >= 1_000 {
            format!("{:.1}k", n as f64 / 1_000.0)
        } else {
            n.to_string()
        }
    }

    fn search(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/search/advanced?order=desc&sort=relevance&site=stackoverflow&q={}&pagesize={}&filter=withbody",
            STACKEXCHANGE_API,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Err(format!("search failed: {}", response.status()));
        }

        let data: SearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .items
            .unwrap_or_default()
            .into_iter()
            .map(|q| {
                let views = Self::format_count(q.view_count);
                let answered_icon = if q.is_answered { "[+]" } else { "[-]" };

                let mut lines = Vec::new();
                lines.push(format!(
                    "Score: {} | Answers: {} {} | Views: {}",
                    q.score, q.answer_count, answered_icon, views
                ));
                if let Some(tags) = &q.tags {
                    if !tags.is_empty() {
                        lines.push(format!(
                            "Tags: {}",
                            tags.iter().take(5).cloned().collect::<Vec<_>>().join(", ")
                        ));
                    }
                }
                if let Some(owner) = &q.owner {
                    if let Some(name) = &owner.display_name {
                        let rep = owner
                            .reputation
                            .map(|r| format!(" ({})", Self::format_count(r)))
                            .unwrap_or_default();
                        lines.push(format!("Asked by: {}{}", name, rep));
                    }
                }

                let mut metadata = HashMap::new();
                metadata.insert("questionId".to_string(), serde_json::json!(q.question_id));
                metadata.insert("score".to_string(), serde_json::json!(q.score));
                metadata.insert("answerCount".to_string(), serde_json::json!(q.answer_count));
                metadata.insert("isAnswered".to_string(), serde_json::json!(q.is_answered));
                metadata.insert("viewCount".to_string(), serde_json::json!(q.view_count));
                if let Some(tags) = &q.tags {
                    metadata.insert("tags".to_string(), serde_json::json!(tags));
                }

                KnowledgeEntry {
                    title: q.title,
                    summary: lines.join("\n"),
                    url: Some(q.link),
                    source: "stackoverflow".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }
}

impl Default for StackOverflowProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for StackOverflowProvider {
    fn name(&self) -> &'static str {
        "stackoverflow"
    }

    fn is_available(&self) -> bool {
        true
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        match self.search(query, limit) {
            Ok(entries) => LookupResult::success(self.name(), entries),
            Err(e) => LookupResult::error(self.name(), e),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn stackoverflow_search() {
        let provider = StackOverflowProvider::new();
        let result = provider.lookup("rust ownership", &LookupOptions::default());
        assert!(result.success);
        assert!(!result.entries.is_empty());
    }
}
